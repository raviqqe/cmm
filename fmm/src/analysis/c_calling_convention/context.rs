use crate::build::NameGenerator;
use std::{cell::RefCell, rc::Rc};

pub struct Context {
    name_generator: Rc<RefCell<NameGenerator>>,
    word_bytes: usize,
}

impl Context {
    pub fn new(word_bytes: usize) -> Self {
        Self {
            name_generator: Rc::new(NameGenerator::new("_c_").into()),
            word_bytes,
        }
    }

    pub fn name_generator(&self) -> Rc<RefCell<NameGenerator>> {
        self.name_generator.clone()
    }

    pub fn word_bytes(&self) -> usize {
        self.word_bytes
    }
}
