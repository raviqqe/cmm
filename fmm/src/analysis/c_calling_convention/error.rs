use crate::{analysis::type_check::TypeCheckError, build::BuildError};
use std::{
    error::Error,
    fmt::{self, Display, Formatter},
};

#[derive(Clone, Debug, PartialEq)]
pub enum CCallingConventionError {
    Build(BuildError),
    TypeCheck(TypeCheckError),
    WordSize(usize),
}

impl Display for CCallingConventionError {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        write!(formatter, "{:?}", self)
    }
}

impl Error for CCallingConventionError {}

impl From<BuildError> for CCallingConventionError {
    fn from(error: BuildError) -> Self {
        Self::Build(error)
    }
}

impl From<TypeCheckError> for CCallingConventionError {
    fn from(error: TypeCheckError) -> Self {
        Self::TypeCheck(error)
    }
}
