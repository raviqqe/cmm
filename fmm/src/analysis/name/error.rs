use std::{
    error::Error,
    fmt::{self, Display, Formatter},
};

#[derive(Clone, Debug, PartialEq)]
pub enum NameError {
    DuplicateNames(String),
}

impl Display for NameError {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        write!(formatter, "{:?}", self)
    }
}

impl Error for NameError {}
