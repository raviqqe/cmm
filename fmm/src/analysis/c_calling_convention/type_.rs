use super::context::Context;
use crate::{
    analysis::type_size,
    types::{self, void_type, Type},
};

pub fn transform_function(context: &Context, function: &types::Function) -> types::Function {
    if function.calling_convention() == types::CallingConvention::Target {
        let is_result_memory = is_memory_class(context, function.result());

        types::Function::new(
            function
                .arguments()
                .iter()
                .map(|type_| transform_type(context, type_))
                .chain(if is_result_memory {
                    Some(transform_type(context, function.result()).into())
                } else {
                    None
                })
                .collect(),
            if is_result_memory {
                void_type().into()
            } else {
                function.result().clone()
            },
            function.calling_convention(),
        )
    } else {
        function.clone()
    }
}

fn transform_type(context: &Context, type_: &Type) -> Type {
    if is_memory_class(context, type_) {
        types::Pointer::new(type_.clone()).into()
    } else {
        type_.clone()
    }
}

// The name, "memory class" comes from the C ABI on System V.
pub fn is_memory_class(context: &Context, type_: &Type) -> bool {
    match type_ {
        Type::Record(record) => {
            type_size::calculate_size(type_, context.word_bytes()) > 2 * context.word_bytes()
                || record
                    .fields()
                    .iter()
                    .any(|type_| is_memory_class(context, type_))
        }
        Type::Union(union) => union
            .members()
            .iter()
            .any(|type_| is_memory_class(context, type_)),
        Type::Function(_) | Type::Pointer(_) | Type::Primitive(_) => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const WORD_BYTES: usize = 8;

    mod function {
        use super::*;
        use pretty_assertions::assert_eq;

        #[test]
        fn transform_function_of_source_calling_convention() {
            let record = types::Record::new(vec![
                types::Primitive::Integer64.into(),
                types::Primitive::Integer64.into(),
                types::Primitive::Integer64.into(),
            ]);
            let function =
                types::Function::new(vec![], record.clone(), types::CallingConvention::Source);

            assert_eq!(
                transform_function(&Context::new(WORD_BYTES), &function),
                function
            );
        }

        #[test]
        fn transform_compatible_function() {
            let function = types::Function::new(
                vec![],
                types::Primitive::Integer64,
                types::CallingConvention::Target,
            );

            assert_eq!(
                transform_function(&Context::new(WORD_BYTES), &function),
                function
            );
        }

        #[test]
        fn transform_function_argument() {
            let record = types::Record::new(vec![
                types::Primitive::Integer64.into(),
                types::Primitive::Integer64.into(),
                types::Primitive::Integer64.into(),
            ]);
            let function = types::Function::new(
                vec![record.clone().into()],
                void_type(),
                types::CallingConvention::Target,
            );

            assert_eq!(
                transform_function(&Context::new(WORD_BYTES), &function),
                types::Function::new(
                    vec![types::Pointer::new(record).into()],
                    void_type(),
                    types::CallingConvention::Target
                )
            );
        }

        #[test]
        fn transform_function_result() {
            let record = types::Record::new(vec![
                types::Primitive::Integer64.into(),
                types::Primitive::Integer64.into(),
                types::Primitive::Integer64.into(),
            ]);
            let function =
                types::Function::new(vec![], record.clone(), types::CallingConvention::Target);

            assert_eq!(
                transform_function(&Context::new(WORD_BYTES), &function),
                types::Function::new(
                    vec![types::Pointer::new(record).into()],
                    void_type(),
                    types::CallingConvention::Target
                )
            );
        }
    }

    mod memory_class {
        use super::*;

        #[test]
        fn function() {
            assert!(!is_memory_class(
                &Context::new(WORD_BYTES),
                &types::Function::new(
                    vec![],
                    types::Primitive::PointerInteger,
                    types::CallingConvention::Target
                )
                .into()
            ));
        }

        #[test]
        fn primitive() {
            assert!(!is_memory_class(
                &Context::new(WORD_BYTES),
                &types::Primitive::PointerInteger.into()
            ));
        }

        #[test]
        fn pointer() {
            assert!(!is_memory_class(
                &Context::new(WORD_BYTES),
                &types::Pointer::new(types::Primitive::PointerInteger).into()
            ));
        }

        #[test]
        fn empty_record() {
            assert!(!is_memory_class(
                &Context::new(WORD_BYTES),
                &types::Record::new(vec![]).into()
            ));
        }

        #[test]
        fn small_record() {
            assert!(!is_memory_class(
                &Context::new(WORD_BYTES),
                &types::Record::new(vec![
                    types::Primitive::Integer64.into(),
                    types::Primitive::Integer64.into()
                ])
                .into()
            ));
        }

        #[test]
        fn large_record() {
            assert!(is_memory_class(
                &Context::new(WORD_BYTES),
                &types::Record::new(vec![
                    types::Primitive::Integer64.into(),
                    types::Primitive::Integer64.into(),
                    types::Primitive::Integer64.into()
                ])
                .into()
            ));
        }
    }
}
