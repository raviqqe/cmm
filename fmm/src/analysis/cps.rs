mod context;
mod continuation_type;
mod error;
mod free_variable;
mod function_type;
mod if_;
mod source_function;
mod stack;
mod target_function;

use self::context::CpsContext;
use super::{name, type_check};
use crate::{ir::*, types::Type};
use error::CpsError;

pub fn transform(module: &Module, result_type: impl Into<Type>) -> Result<Module, CpsError> {
    name::check(module)?;
    type_check::check(module)?;

    let context = CpsContext::new(result_type.into());

    let module = if_::flatten(module);
    let module = source_function::transform(&context, &module)?;
    let module = target_function::transform(&context, &module)?;
    let module = function_type::transform(&module, context.result_type());

    name::check(&module)?;
    type_check::check(&module)?;

    Ok(module)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        analysis::format,
        types::{self, void_type, CallingConvention, Type},
    };
    use stack::type_;

    fn create_function_type(arguments: Vec<Type>, result: impl Into<Type>) -> types::Function {
        types::Function::new(arguments, result, CallingConvention::Source)
    }

    fn create_cps_function_type(arguments: Vec<Type>, result: impl Into<Type>) -> types::Function {
        types::Function::new(
            [
                type_(),
                types::Function::new(
                    vec![type_(), result.into()],
                    void_type(),
                    CallingConvention::Tail,
                )
                .into(),
            ]
            .into_iter()
            .chain(arguments)
            .collect(),
            void_type(),
            CallingConvention::Tail,
        )
    }

    fn create_function_definition(
        name: impl Into<String>,
        arguments: Vec<Argument>,
        result_type: impl Into<Type>,
        body: Block,
    ) -> FunctionDefinition {
        FunctionDefinition::new(name, arguments, result_type, body, Default::default())
    }

    fn test_transformation(module: &Module) {
        let one = transform(module, void_type()).unwrap();
        let other = transform(module, void_type()).unwrap();

        name::check(&one).unwrap();
        type_check::check(&one).unwrap();

        assert_eq!(one, other);
    }

    #[test]
    fn transform_empty_module() {
        test_transformation(&Module::new(vec![], vec![], vec![], vec![]));
    }

    #[test]
    fn transform_function_definition() {
        test_transformation(&Module::new(
            vec![],
            vec![],
            vec![],
            vec![create_function_definition(
                "f",
                vec![],
                types::Primitive::Float64,
                Block::new(
                    vec![],
                    Return::new(types::Primitive::Float64, Primitive::Float64(42.0)),
                ),
            )],
        ));
    }

    #[test]
    fn transform_call() {
        let function_type = create_function_type(
            vec![types::Primitive::Float64.into()],
            types::Primitive::Float64,
        );

        test_transformation(&Module::new(
            vec![],
            vec![FunctionDeclaration::new("f", function_type.clone())],
            vec![],
            vec![create_function_definition(
                "g",
                vec![],
                types::Primitive::Float64,
                Block::new(
                    vec![Call::new(
                        function_type,
                        Variable::new("f"),
                        vec![Primitive::Float64(42.0).into()],
                        "x",
                    )
                    .into()],
                    Return::new(types::Primitive::Float64, Variable::new("x")),
                ),
            )],
        ));
    }

    #[test]
    fn transform_instruction_after_call() {
        let function_type =
            create_function_type(vec![], types::Pointer::new(types::Primitive::Float64));

        test_transformation(&Module::new(
            vec![],
            vec![FunctionDeclaration::new("f", function_type.clone())],
            vec![],
            vec![create_function_definition(
                "g",
                vec![],
                types::Primitive::Float64,
                Block::new(
                    vec![
                        Call::new(function_type, Variable::new("f"), vec![], "x").into(),
                        Load::new(types::Primitive::Float64, Variable::new("x"), "y").into(),
                    ],
                    Return::new(types::Primitive::Float64, Variable::new("y")),
                ),
            )],
        ));
    }

    #[test]
    fn transform_two_calls() {
        let function_type = create_function_type(
            vec![types::Primitive::PointerInteger.into()],
            types::Primitive::PointerInteger,
        );

        test_transformation(&Module::new(
            vec![],
            vec![FunctionDeclaration::new("f", function_type.clone())],
            vec![],
            vec![create_function_definition(
                "g",
                vec![],
                types::Primitive::PointerInteger,
                Block::new(
                    vec![
                        Call::new(
                            function_type.clone(),
                            Variable::new("f"),
                            vec![Primitive::PointerInteger(42).into()],
                            "x",
                        )
                        .into(),
                        Call::new(
                            function_type,
                            Variable::new("f"),
                            vec![Variable::new("x").into()],
                            "y",
                        )
                        .into(),
                    ],
                    Return::new(types::Primitive::PointerInteger, Variable::new("y")),
                ),
            )],
        ));
    }

    #[test]
    fn transform_tail_call() {
        let function_type = create_function_type(
            vec![types::Primitive::Float64.into()],
            types::Primitive::Float64,
        );
        let cps_function_type = create_cps_function_type(
            vec![types::Primitive::Float64.into()],
            types::Primitive::Float64,
        );

        pretty_assertions::assert_eq!(
            transform(
                &Module::new(
                    vec![],
                    vec![FunctionDeclaration::new("f", function_type.clone())],
                    vec![],
                    vec![create_function_definition(
                        "g",
                        vec![],
                        types::Primitive::Float64,
                        Block::new(
                            vec![Call::new(
                                function_type,
                                Variable::new("f"),
                                vec![Primitive::Float64(42.0).into()],
                                "x",
                            )
                            .into()],
                            Return::new(types::Primitive::Float64, Variable::new("x")),
                        ),
                    )],
                ),
                void_type()
            ),
            Ok(Module::new(
                vec![],
                vec![FunctionDeclaration::new("f", cps_function_type.clone())],
                vec![],
                vec![FunctionDefinition::new(
                    "g",
                    vec![
                        Argument::new("_s", type_()),
                        Argument::new(
                            "_k",
                            types::Function::new(
                                vec![type_(), types::Primitive::Float64.into()],
                                void_type(),
                                CallingConvention::Tail,
                            )
                        ),
                    ],
                    void_type(),
                    Block::new(
                        vec![Call::new(
                            cps_function_type,
                            Variable::new("f"),
                            vec![
                                Variable::new("_s").into(),
                                Variable::new("_k").into(),
                                Primitive::Float64(42.0).into()
                            ],
                            "_k_0",
                        )
                        .into()],
                        Return::new(void_type(), Variable::new("_k_0")),
                    ),
                    FunctionDefinitionOptions::new()
                        .set_calling_convention(CallingConvention::Tail)
                )],
            ))
        );
    }

    #[test]
    fn transform_call_with_continuation_with_free_variable() {
        let function_type = create_function_type(vec![], types::Primitive::PointerInteger);

        test_transformation(&Module::new(
            vec![],
            vec![FunctionDeclaration::new("f", function_type.clone())],
            vec![],
            vec![create_function_definition(
                "g",
                vec![Argument::new(
                    "p",
                    types::Pointer::new(types::Primitive::PointerInteger),
                )],
                types::Primitive::PointerInteger,
                Block::new(
                    vec![
                        Load::new(types::Primitive::PointerInteger, Variable::new("p"), "x").into(),
                        Call::new(function_type, Variable::new("f"), vec![], "y").into(),
                        Store::new(
                            types::Primitive::PointerInteger,
                            Variable::new("x"),
                            Variable::new("p"),
                        )
                        .into(),
                        Store::new(
                            types::Primitive::PointerInteger,
                            Variable::new("y"),
                            Variable::new("p"),
                        )
                        .into(),
                    ],
                    Return::new(
                        types::Primitive::PointerInteger,
                        ArithmeticOperation::new(
                            types::Primitive::PointerInteger,
                            ArithmeticOperator::Add,
                            Variable::new("x"),
                            Variable::new("y"),
                        ),
                    ),
                ),
            )],
        ));
    }

    #[test]
    fn transform_call_with_continuation_with_free_variable_from_if() {
        let function_type = create_function_type(vec![], types::Primitive::PointerInteger);
        let pointer_type = types::Pointer::new(types::Primitive::PointerInteger);

        test_transformation(&Module::new(
            vec![],
            vec![FunctionDeclaration::new("f", function_type.clone())],
            vec![],
            vec![create_function_definition(
                "g",
                vec![Argument::new(
                    "p",
                    types::Pointer::new(types::Primitive::PointerInteger),
                )],
                types::Primitive::PointerInteger,
                Block::new(
                    vec![
                        Load::new(types::Primitive::PointerInteger, Variable::new("p"), "x").into(),
                        If::new(
                            pointer_type.clone(),
                            Primitive::Boolean(true),
                            Block::new(vec![], Branch::new(pointer_type, Variable::new("p"))),
                            Block::new(vec![], TerminalInstruction::Unreachable),
                            "q",
                        )
                        .into(),
                        Call::new(function_type, Variable::new("f"), vec![], "y").into(),
                        Store::new(
                            types::Primitive::PointerInteger,
                            Variable::new("x"),
                            Variable::new("p"),
                        )
                        .into(),
                        Store::new(
                            types::Primitive::PointerInteger,
                            Variable::new("y"),
                            Variable::new("q"),
                        )
                        .into(),
                    ],
                    Return::new(
                        types::Primitive::PointerInteger,
                        ArithmeticOperation::new(
                            types::Primitive::PointerInteger,
                            ArithmeticOperator::Add,
                            Variable::new("x"),
                            Variable::new("y"),
                        ),
                    ),
                ),
            )],
        ));
    }

    mod if_ {
        use super::*;

        #[test]
        fn transform_call_in_block() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("f", function_type.clone())],
                vec![],
                vec![create_function_definition(
                    "g",
                    vec![],
                    types::Primitive::Float64,
                    Block::new(
                        vec![If::new(
                            types::Primitive::Float64,
                            Primitive::Boolean(true),
                            Block::new(
                                vec![Call::new(
                                    function_type,
                                    Variable::new("f"),
                                    vec![Primitive::Float64(42.0).into()],
                                    "x",
                                )
                                .into()],
                                Branch::new(types::Primitive::Float64, Variable::new("x")),
                            ),
                            Block::new(vec![], TerminalInstruction::Unreachable),
                            "y",
                        )
                        .into()],
                        Return::new(types::Primitive::Float64, Variable::new("y")),
                    ),
                )],
            ));
        }

        #[test]
        fn transform_call_before_branch_into_tail_call() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            insta::assert_snapshot!(format::format_module(
                &transform(
                    &Module::new(
                        vec![],
                        vec![FunctionDeclaration::new("f", function_type.clone())],
                        vec![],
                        vec![create_function_definition(
                            "g",
                            vec![],
                            types::Primitive::Float64,
                            Block::new(
                                vec![If::new(
                                    types::Primitive::Float64,
                                    Primitive::Boolean(true),
                                    Block::new(
                                        vec![Call::new(
                                            function_type,
                                            Variable::new("f"),
                                            vec![Primitive::Float64(42.0).into()],
                                            "x",
                                        )
                                        .into()],
                                        Branch::new(types::Primitive::Float64, Variable::new("x")),
                                    ),
                                    Block::new(vec![], TerminalInstruction::Unreachable),
                                    "y",
                                )
                                .into()],
                                Return::new(types::Primitive::Float64, Variable::new("y")),
                            ),
                        )],
                    ),
                    void_type()
                )
                .unwrap()
            ));
        }

        #[test]
        fn transform_tail_call_in_block() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("f", function_type.clone())],
                vec![],
                vec![create_function_definition(
                    "g",
                    vec![],
                    types::Primitive::Float64,
                    Block::new(
                        vec![If::new(
                            void_type(),
                            Primitive::Boolean(true),
                            Block::new(
                                vec![Call::new(
                                    function_type,
                                    Variable::new("f"),
                                    vec![Primitive::Float64(42.0).into()],
                                    "x",
                                )
                                .into()],
                                Return::new(types::Primitive::Float64, Variable::new("x")),
                            ),
                            Block::new(vec![], TerminalInstruction::Unreachable),
                            "_",
                        )
                        .into()],
                        TerminalInstruction::Unreachable,
                    ),
                )],
            ));
        }

        #[test]
        fn transform_non_tail_call_in_block() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            insta::assert_snapshot!(format::format_module(
                &transform(
                    &Module::new(
                        vec![],
                        vec![FunctionDeclaration::new("f", function_type.clone())],
                        vec![],
                        vec![create_function_definition(
                            "g",
                            vec![],
                            types::Primitive::Float64,
                            Block::new(
                                vec![If::new(
                                    types::Primitive::Float64,
                                    Primitive::Boolean(true),
                                    Block::new(
                                        vec![Call::new(
                                            function_type,
                                            Variable::new("f"),
                                            vec![Primitive::Float64(42.0).into()],
                                            "x",
                                        )
                                        .into()],
                                        Branch::new(types::Primitive::Float64, Variable::new("x")),
                                    ),
                                    Block::new(vec![], TerminalInstruction::Unreachable),
                                    "y",
                                )
                                .into()],
                                Return::new(
                                    types::Primitive::Float64,
                                    ArithmeticOperation::new(
                                        types::Primitive::Float64,
                                        ArithmeticOperator::Add,
                                        Variable::new("y"),
                                        Variable::new("y")
                                    )
                                ),
                            ),
                        )],
                    ),
                    void_type()
                )
                .unwrap()
            ));
        }

        #[test]
        fn transform_call_in_block_and_call_after_if() {
            let function_type = create_function_type(
                vec![types::Primitive::PointerInteger.into()],
                types::Primitive::PointerInteger,
            );

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("f", function_type.clone())],
                vec![],
                vec![create_function_definition(
                    "g",
                    vec![],
                    types::Primitive::PointerInteger,
                    Block::new(
                        vec![
                            If::new(
                                types::Primitive::PointerInteger,
                                Primitive::Boolean(true),
                                Block::new(
                                    vec![Call::new(
                                        function_type.clone(),
                                        Variable::new("f"),
                                        vec![Primitive::PointerInteger(42).into()],
                                        "x",
                                    )
                                    .into()],
                                    Branch::new(
                                        types::Primitive::PointerInteger,
                                        Variable::new("x"),
                                    ),
                                ),
                                Block::new(vec![], TerminalInstruction::Unreachable),
                                "y",
                            )
                            .into(),
                            Call::new(
                                function_type,
                                Variable::new("f"),
                                vec![Variable::new("y").into()],
                                "z",
                            )
                            .into(),
                        ],
                        Return::new(types::Primitive::PointerInteger, Variable::new("z")),
                    ),
                )],
            ));
        }

        #[test]
        fn transform_call_after_preserved_if() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            insta::assert_snapshot!(format::format_module(
                &transform(
                    &Module::new(
                        vec![],
                        vec![FunctionDeclaration::new("f", function_type.clone())],
                        vec![],
                        vec![create_function_definition(
                            "g",
                            vec![],
                            types::Primitive::Float64,
                            Block::new(
                                vec![
                                    If::new(
                                        types::Primitive::Float64,
                                        Primitive::Boolean(true),
                                        Block::new(
                                            vec![Load::new(
                                                types::Primitive::Float64,
                                                Undefined::new(types::Pointer::new(
                                                    types::Primitive::Float64
                                                )),
                                                "x"
                                            )
                                            .into()],
                                            Branch::new(
                                                types::Primitive::Float64,
                                                Variable::new("x")
                                            ),
                                        ),
                                        Block::new(vec![], TerminalInstruction::Unreachable),
                                        "y",
                                    )
                                    .into(),
                                    Call::new(
                                        function_type,
                                        Variable::new("f"),
                                        vec![Variable::new("y").into()],
                                        "z",
                                    )
                                    .into()
                                ],
                                Return::new(types::Primitive::Float64, Variable::new("z")),
                            ),
                        )],
                    ),
                    void_type()
                )
                .unwrap()
            ));
        }

        #[test]
        fn transform_call_in_block_with_multiple_free_variables() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("f", function_type.clone())],
                vec![],
                vec![create_function_definition(
                    "g",
                    vec![
                        Argument::new("a", types::Primitive::Float64),
                        Argument::new("b", types::Primitive::Float64),
                    ],
                    types::Primitive::Float64,
                    Block::new(
                        vec![If::new(
                            types::Primitive::Float64,
                            Primitive::Boolean(true),
                            Block::new(
                                vec![Call::new(
                                    function_type,
                                    Variable::new("f"),
                                    vec![Primitive::Float64(42.0).into()],
                                    "x",
                                )
                                .into()],
                                Branch::new(types::Primitive::Float64, Variable::new("x")),
                            ),
                            Block::new(vec![], TerminalInstruction::Unreachable),
                            "y",
                        )
                        .into()],
                        Return::new(
                            types::Primitive::Float64,
                            ArithmeticOperation::new(
                                types::Primitive::Float64,
                                ArithmeticOperator::Add,
                                ArithmeticOperation::new(
                                    types::Primitive::Float64,
                                    ArithmeticOperator::Add,
                                    Variable::new("a"),
                                    Variable::new("b"),
                                ),
                                Variable::new("y"),
                            ),
                        ),
                    ),
                )],
            ));
        }

        #[test]
        fn transform_call_in_block_with_continuation_with_instructions() {
            let function_type = create_function_type(vec![], types::Primitive::PointerInteger);

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("f", function_type.clone())],
                vec![],
                vec![create_function_definition(
                    "g",
                    vec![Argument::new(
                        "p",
                        types::Pointer::new(types::Primitive::PointerInteger),
                    )],
                    types::Primitive::PointerInteger,
                    Block::new(
                        vec![
                            Load::new(types::Primitive::PointerInteger, Variable::new("p"), "x")
                                .into(),
                            If::new(
                                types::Primitive::PointerInteger,
                                Primitive::Boolean(true),
                                Block::new(
                                    vec![Call::new(function_type, Variable::new("f"), vec![], "a")
                                        .into()],
                                    Branch::new(
                                        types::Primitive::PointerInteger,
                                        Variable::new("a"),
                                    ),
                                ),
                                Block::new(
                                    vec![],
                                    Branch::new(
                                        types::Primitive::PointerInteger,
                                        Primitive::PointerInteger(2),
                                    ),
                                ),
                                "y",
                            )
                            .into(),
                            Store::new(
                                types::Primitive::PointerInteger,
                                Variable::new("x"),
                                Variable::new("p"),
                            )
                            .into(),
                            Store::new(
                                types::Primitive::PointerInteger,
                                Variable::new("y"),
                                Variable::new("p"),
                            )
                            .into(),
                        ],
                        Return::new(
                            types::Primitive::PointerInteger,
                            ArithmeticOperation::new(
                                types::Primitive::PointerInteger,
                                ArithmeticOperator::Add,
                                Variable::new("x"),
                                Variable::new("y"),
                            ),
                        ),
                    ),
                )],
            ));
        }

        #[test]
        fn transform_if_with_source_call_and_preserved_if() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("f", function_type.clone())],
                vec![],
                vec![create_function_definition(
                    "g",
                    vec![],
                    types::Primitive::Float64,
                    Block::new(
                        vec![If::new(
                            types::Primitive::Float64,
                            Primitive::Boolean(true),
                            Block::new(
                                vec![
                                    If::new(
                                        types::Primitive::Float64,
                                        Primitive::Boolean(true),
                                        Block::new(vec![], TerminalInstruction::Unreachable),
                                        Block::new(vec![], TerminalInstruction::Unreachable),
                                        "",
                                    )
                                    .into(),
                                    Call::new(
                                        function_type,
                                        Variable::new("f"),
                                        vec![Primitive::Float64(42.0).into()],
                                        "x",
                                    )
                                    .into(),
                                ],
                                Branch::new(types::Primitive::Float64, Variable::new("x")),
                            ),
                            Block::new(vec![], TerminalInstruction::Unreachable),
                            "y",
                        )
                        .into()],
                        Return::new(types::Primitive::Float64, Variable::new("y")),
                    ),
                )],
            ));
        }
    }

    mod target_function_definition {
        use super::*;

        #[test]
        fn transform_with_no_argument() {
            let function_type = create_function_type(vec![], types::Primitive::Float64);

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("g", function_type.clone())],
                vec![],
                vec![FunctionDefinition::new(
                    "f",
                    vec![],
                    types::Primitive::Float64,
                    Block::new(
                        vec![Call::new(function_type, Variable::new("g"), vec![], "x").into()],
                        Return::new(types::Primitive::Float64, Variable::new("x")),
                    ),
                    FunctionDefinitionOptions::new()
                        .set_calling_convention(CallingConvention::Target),
                )],
            ));
        }

        #[test]
        fn transform_one_argument() {
            let function_type = create_function_type(
                vec![types::Primitive::Float64.into()],
                types::Primitive::Float64,
            );

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("g", function_type.clone())],
                vec![],
                vec![FunctionDefinition::new(
                    "f",
                    vec![],
                    types::Primitive::Float64,
                    Block::new(
                        vec![Call::new(
                            function_type,
                            Variable::new("g"),
                            vec![Primitive::Float64(42.0).into()],
                            "x",
                        )
                        .into()],
                        Return::new(types::Primitive::Float64, Variable::new("x")),
                    ),
                    FunctionDefinitionOptions::new()
                        .set_calling_convention(CallingConvention::Target),
                )],
            ));
        }

        #[test]
        fn transform_two_argument() {
            let function_type = create_function_type(
                vec![
                    types::Primitive::Float64.into(),
                    types::Primitive::Integer64.into(),
                ],
                types::Primitive::Float64,
            );

            test_transformation(&Module::new(
                vec![],
                vec![FunctionDeclaration::new("g", function_type.clone())],
                vec![],
                vec![FunctionDefinition::new(
                    "f",
                    vec![],
                    types::Primitive::Float64,
                    Block::new(
                        vec![Call::new(
                            function_type,
                            Variable::new("g"),
                            vec![
                                Primitive::Float64(42.0).into(),
                                Primitive::Integer64(42).into(),
                            ],
                            "x",
                        )
                        .into()],
                        Return::new(types::Primitive::Float64, Variable::new("x")),
                    ),
                    FunctionDefinitionOptions::new()
                        .set_calling_convention(CallingConvention::Target),
                )],
            ));
        }
    }
}
