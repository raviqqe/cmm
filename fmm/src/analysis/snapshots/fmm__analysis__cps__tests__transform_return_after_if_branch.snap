---
source: fmm/src/analysis/cps.rs
assertion_line: 305
expression: "format_module(&transform_to_cps(&Module::new(vec![],\n                                             vec![FunctionDeclaration ::\n                                                  new(\"f\",\n                                                      function_type.clone())],\n                                             vec![],\n                                             vec![create_function_definition(\"g\",\n                                                                             vec!\n                                                                             [],\n                                                                             Block\n                                                                             ::\n                                                                             new(vec!\n                                                                                 [If\n                                                                                  ::\n                                                                                  new(types\n                                                                                      ::\n                                                                                      Primitive\n                                                                                      ::\n                                                                                      Float64,\n                                                                                      Primitive\n                                                                                      ::\n                                                                                      Boolean(true),\n                                                                                      Block\n                                                                                      ::\n                                                                                      new(vec!\n                                                                                          [Call\n                                                                                           ::\n                                                                                           new(function_type,\n                                                                                               Variable\n                                                                                               ::\n                                                                                               new(\"f\"),\n                                                                                               vec!\n                                                                                               [Primitive\n                                                                                                ::\n                                                                                                Float64(42.0).into()],\n                                                                                               \"x\",).into()],\n                                                                                          Branch\n                                                                                          ::\n                                                                                          new(types\n                                                                                              ::\n                                                                                              Primitive\n                                                                                              ::\n                                                                                              Float64,\n                                                                                              Variable\n                                                                                              ::\n                                                                                              new(\"x\")),),\n                                                                                      Block\n                                                                                      ::\n                                                                                      new(vec!\n                                                                                          [],\n                                                                                          TerminalInstruction\n                                                                                          ::\n                                                                                          Unreachable),\n                                                                                      \"y\",).into()],\n                                                                                 Return\n                                                                                 ::\n                                                                                 new(types\n                                                                                     ::\n                                                                                     Primitive\n                                                                                     ::\n                                                                                     Float64,\n                                                                                     ArithmeticOperation\n                                                                                     ::\n                                                                                     new(types\n                                                                                         ::\n                                                                                         Primitive\n                                                                                         ::\n                                                                                         Float64,\n                                                                                         ArithmeticOperator\n                                                                                         ::\n                                                                                         Add,\n                                                                                         Variable\n                                                                                         ::\n                                                                                         new(\"y\"),\n                                                                                         Variable\n                                                                                         ::\n                                                                                         new(\"y\"))),),\n                                                                             types\n                                                                             ::\n                                                                             Primitive\n                                                                             ::\n                                                                             Float64,)]),\n                                VOID_TYPE.clone()).unwrap())"

---
(module
  (function g _s _k
    (block
      (if true
        (block
          (load (record-address _s 1) _k_6)
          (if (== (size-of record) 0)
            (block
              (branch 0))
            (block
              (branch (* (+ (/ (- (size-of record) 1) (align-of pointer-integer)) 1) (align-of pointer-integer))))
            _k_7)
          (load (record-address _s 2) _k_8)
          (if (> (+ _k_6 _k_7) _k_8)
            (block
              (load (record-address _s 0) _k_9)
              (reallocate-heap _k_9 (* _k_8 2) _k_a)
              (store _k_a (record-address _s 0))
              (store (* _k_8 2) (record-address _s 2))
              (branch (record)))
            (block
              (branch (record)))
            _k_b)
          (load (record-address _s 0) _k_c)
          (load (record-address _s 1) _k_d)
          (store (record _k) (bit-cast (pointer-address _k_c _k_d)))
          (store (+ _k_6 _k_7) (record-address _s 1))
          (call f _s _k_0 42 _result)
          (return _result))
        (block
          (unreachable))
        _k_e)
      (unreachable)))
  (function _k_0 _s x
    (block
      (load (record-address _s 1) _k_1)
      (if (== (size-of record) 0)
        (block
          (branch 0))
        (block
          (branch (* (+ (/ (- (size-of record) 1) (align-of pointer-integer)) 1) (align-of pointer-integer))))
        _k_2)
      (store (- _k_1 _k_2) (record-address _s 1))
      (load (record-address _s 0) _k_3)
      (load (record-address _s 1) _k_4)
      (load (bit-cast (pointer-address _k_3 _k_4)) _k_5)
      (deconstruct-record _k_5 0 _k)
      (call _k _s (+ x x) _result)
      (return _result))))
